tests/integration-tests/tests/common/scaling_common.py

```diff
@@ -16,12 +18,113 @@
 from remote_command_executor import RemoteCommandExecutor
 from retrying import RetryError, retry
 from time_utils import seconds
-from utils import get_compute_nodes_count
+from utils import CWMetric, get_compute_nodes_instance_count, publish_metrics_to_cloudwatch
 
 SCALING_COMMON_DATADIR = pathlib.Path(__file__).parent / "scaling"
 
 
-def get_compute_nodes_allocation(scheduler_commands, region, stack_name, max_monitoring_time):
+def scaling_target_condition(
+    ec2_capacity_time_series,
+    compute_nodes_time_series,
+    target_cluster_size,
+    use_ec2_limit=True,  # Stop monitoring after all EC2 instances have been launched
+    use_compute_nodes_limit=True,  # Stop monitoring after all nodes have joined the cluster
+):
+    return (
+        (use_ec2_limit and ec2_capacity_time_series[-1] != target_cluster_size)
+        or (use_compute_nodes_limit and compute_nodes_time_series[-1] != target_cluster_size)
+        or (use_ec2_limit and max(ec2_capacity_time_series) == 0)
+        or (use_compute_nodes_limit and max(compute_nodes_time_series) == 0)
+    )
+
+
+def get_scaling_metrics(
+    remote_command_executor: RemoteCommandExecutor,
+    region,
+    cluster_name,
+    max_monitoring_time,
+    target_cluster_size=0,
+    publish_metrics=False,
+):
+    ec2_capacity_time_series = []
+    compute_nodes_time_series = []
+    timestamps = []
+    cw_client = boto3.client("cloudwatch", region_name=region)
+
+    @retry(
+        # Retry until EC2 and Scheduler capacities scale to specified target
+        retry_on_result=lambda _: scaling_target_condition(
+            ec2_capacity_time_series, compute_nodes_time_series, target_cluster_size
+        ),
+        wait_fixed=seconds(1),
+        stop_max_delay=max_monitoring_time,
+    )
+    def _collect_metrics():
+        remote_command_executor.run_remote_script(
+            script_file=str(SCALING_COMMON_DATADIR / "collect_metrics_on_headnode.sh")
+        )
+        headnode_metrics = json.loads(
+            remote_command_executor.run_remote_command(command="cat $HOME/scaling_metrics.json").stdout
+        )
+        logging.info(f"Metrics from HeadNode: {headnode_metrics}")
+        compute_node_count = int(headnode_metrics.get("NodeCount"))
+        pending_jobs_count = int(headnode_metrics.get("PendingJobsCount"))
```

Comment:
minor: some of these variables are useful only in the `if publish_metrics` branch.

Reply:
fixed